service/storage_service.cc

```diff
@@ -1181,116 +1171,169 @@ class topology_coordinator {
+                               std::function<future<>()> action) {
+        if (!holder || holder->failed()) {
+            holder = futurize_invoke(action)
+                        .finally([this, g = _async_gate.hold(), gid, name] () noexcept {
+                slogger.trace("raft topology: {} for tablet {} resolved.", name, gid);
+                _tablets_ready = true;
+                _topo_sm.event.broadcast();
+            });
+            return false;
         }
+
+        if (!holder->available()) {
+            slogger.trace("raft topology: Tablet {} still doing {}", gid, name);
+            return false;
+        }
+
+        return true;
     }
 
-    future<> del_tablet_transitions(std::vector<canonical_mutation>& out, group0_guard& guard) {
+    future<> for_each_tablet_transition(std::function<void(const locator::tablet_map&,
+                                                           schema_ptr,
+                                                           locator::global_tablet_id,
+                                                           const locator::tablet_transition_info&)> func) {
         auto tm = get_token_metadata_ptr();
         for (auto&& [table, tmap] : tm->tablets().all_tables()) {
             co_await coroutine::maybe_yield();
             auto s = _db.find_schema(table);
-            for (auto&& [tablet, trinfo] : tmap.transitions()) {
+            for (auto&& [tablet, trinfo]: tmap.transitions()) {
                 co_await coroutine::maybe_yield();
-                auto last_token = tmap.get_last_token(tablet);
-                out.emplace_back(
-                    replica::tablet_mutation_builder(guard.write_timestamp(), s->ks_name(), table)
-                        .del_transition(last_token)
-                        .set_replicas(last_token, trinfo.next)
-                        .build());
+                auto gid = locator::global_tablet_id {table, tablet};
+                func(tmap, s, gid, trinfo);
             }
         }
     }
 
-    future<> transition_tablets_to_stage(group0_guard guard, locator::tablet_transition_stage stage, topology::transition_state topo_state) {
-        std::vector<canonical_mutation> updates;
-        co_await set_tablet_transition_stage(updates, guard, stage);
-        updates.emplace_back(
-            topology_mutation_builder(guard.write_timestamp())
-                .set_transition_state(topo_state)
-                .set_version(_topo_sm._topology.version + 1)
-                .build());
-        co_await update_topology_state(std::move(guard), std::move(updates), format("Moved tablet migration to stage: {}", stage));
-    }
+    future<> handle_tablet_migration(group0_guard guard) {
+        // This step acts like a pump which advances state machines of individual tablets,
+        // batching barriers and group0 updates.
+        // If progress cannot be made, e.g. because all transitions are streaming, we block
+        // and wait for notification.
```

Comment:
In a way it's similar to the reactor, waiting for futures to complete and launching the next tasks, but here we inject a big wrapper around each task.